Adhere to Same Origin Policy more strictly

cfredric opened the pull request:
This is meant to address the ambiguities pointed out in #210 by defining Storage Access's integration with Fetch more rigorously. This PR codifies the behavior I'd like to Storage Access implementations to align on (namely stricter adherence to the Same Origin Policy).

- [x] At least two implementers are interested (and none opposed):
   * Chrome
   * Webkit
   * Firefox
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://crrev.com/c/6241903 ([wpt.fyi](https://wpt.fyi/results/storage-access-api/requestStorageAccess-cross-origin-fetch.sub.https.tentative.window.html))
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://crbug.com/379030052
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1960071
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=291440


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/cfredric/storage-access/pull/213.html" title="Last updated on Apr 11, 2025, 7:27 PM UTC (1938b57)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/privacycg/storage-access/213/8b80aa2...cfredric:1938b57.html" title="Last updated on Apr 11, 2025, 7:27 PM UTC (1938b57)">Diff</a>

bvandersloot-mozilla commented on storage-access.bs:
```diff
@@ -301,15 +315,33 @@ When creating |request|'s [=reserved client=] in [=create navigation params by f
+This API is intended to be used with environments and user agent configurations that block access to unpartitioned cookies in a [=third party context=]. At the time of this writing, this concept has not yet been integrated into the [=HTTP-network-or-cache fetch=] and {{Document/cookie}} algorithms. To allow for such an integration, the [=cookie store=] will need to be modified to receive informat ...
 
-Once the cookie store allows for receiving information about storage access, we would update [=HTTP-network-or-cache fetch=] and {{Document/cookie}} to pass the [=environment=]'s [=environment/has storage access=] to the [=cookie store=] when retrieving cookies.
+Once the cookie store allows for receiving information about storage access, we would update [=HTTP-network-or-cache fetch=] and {{Document/cookie}} to pass the [=request=]'s [=request/eligible for storage-access=] to the [=cookie store=] when retrieving cookies.
```
I don't think we pass it directly, given what we know now, but yes that is the spirit.

cfredric replied:
I think this algorithm is yet to be defined, right? I.e., updating the prose here is the best I can do at the moment?

bvandersloot-mozilla commented on storage-access.bs:
```diff
@@ -173,6 +175,18 @@ Modify the definition of [=source snapshot params=] in the following manner:
+
+    1. If |request|'s [=request/client=] is null, return false.
+    1. If |request|'s [=request/client=]'s [=environment/has storage access=] is false, return false.
+    1. If |request|'s [=url/origin=] is not [=same origin=] with |request|'s [=request/url=]'s [=url/origin=], return false.
```
So this makes a storage access grant only apply to a single origin, rather than the entire site?

bvandersloot-mozilla replied:
Let me clarify: it makes `document.requestStorageAccess()` only apply to the origin of the document? 

I'd be more comfortable with this if there were a positive signal from Apple around Storage Access Headers, because this further leans on that feature for subresources.

bvandersloot-mozilla replied:
I agree that we should fix the cross-origin-same-site navigation fetch to not send unpartitioned cookies, however the subresource fetch inside of the document is the part that leans on  Storage Access Headers.

cfredric replied:
> Let me clarify: it makes document.requestStorageAccess() only apply to the origin of the document?

`document.requestStorageAccess()` will still request a permission that applies to the whole site, but only the document's origin is opted-in and receives cookie access as a result of the call. Other origins have to explicitly activate the permission by calling `document.requestStorageAccess()` themselves, if they also need cookie access.

> if there were a positive signal from Apple around Storage Access Headers

FWIW, Anne [wrote](https://github.com/WebKit/standards-positions/issues/412#issuecomment-2605170501) that SAH seems like a reasonable extension of Storage Access, and opened one issue about worker support (specifically for the `Activate-Storage-Access: load` header). He didn't mention any concerns about the `Activate-Storage-Access: retry; ...` part, which is what this would lean on. That's not a formal positive signal, though, to be fair.

> I agree that we should fix the cross-origin-same-site navigation fetch to not send unpartitioned cookies, however the subresource fetch inside of the document is the part that leans on Storage Access Headers.

Are you saying that you'd expect SAA's Fetch integration to treat navigation fetches and subresource fetches differently? My assumption was that Fetch would be agnostic to the request's `destination`, so that fetches of different kinds of resources get the same kind of CSRF protection (whether it's a navigation or not); i.e., whether cookies are included wouldn't depend on what exactly the page is fetching. WDYT?

bvandersloot-mozilla replied:
> Other origins have to explicitly activate the permission by calling document.requestStorageAccess() themselves, if they also need cookie access.

So they would need to create another iframe to do that (excepting SAH for subresources)?

> Are you saying that you'd expect SAA's Fetch integration to treat navigation fetches and subresource fetches differently? 

Yeah, that is the way I was thinking. In my mind you are fetching resources that are going into a different context than the current one, so it is natural to treat them as the destination context rather than the source context.

The way I'm thinking about this is like this. The storage access API breaks open the document to use its unpartitioned cookies. An iframe provider controls security by dictating what Origins can request the API.

cfredric replied:
> So they would need to create another iframe to do that (excepting SAH for subresources)?

That's right. The principle is that if an origin didn't opt into cross-site cookie access, then the browser doesn't send cookies to it. And the two ways of opting in are:
* Call `document.requestStorageAccess()` in a document which is same-origin with the server endpoint that wants to receive cookies.
* Use SAH.

> Yeah, that is the way I was thinking. In my mind you are fetching resources that are going into a different context than the current one, so it is natural to treat them as the destination context rather than the source context.

I think that model might be okay from a privacy POV, but not from a web security POV. From a security POV, we need to consider the source of the fetch too.

IMO this is easiest to see if we start by thinking about cross-*site* fetches: we know that when a malicious site can send unfettered credentialed requests to another site, it can carry out [CSRF](https://owasp.org/www-community/attacks/csrf) attacks. Whether the fetch is a navigation, or just a subresource fetch, doesn't matter w.r.t. CSRF.

But the web security model is the Same-Origin Policy, so the web's [security principals](https://en.wikipedia.org/wiki/Principal_(computer_security)) are *origins*, not sites. So we can modify the CSRF attack (`s/site/origin/g`): a malicious origin should not be able to send credentialed requests to another origin (without that origin's prior go-ahead -- CORS' [ACAC](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials) header is prior art here).

This PR aims to achieve that protection -- at least whenever the SAA is the method by which cookies are accessed.

bvandersloot-mozilla reviewed (commented):
It would be nice if the tentative WPT had more coverage of the cases Anne outlined [here](https://github.com/privacycg/storage-access/issues/210#issuecomment-2532000860), although I admit I didn't update them either.

cfredric commented:
Re: the WPT test cases Anne listed, AFAIK the WPT framework [only provides two distinct domains](https://web-platform-tests.org/writing-tests/server-features.html#tests-involving-multiple-origins), so we'd have to add a new one if we wanted to write the tests that involve all of A, B, and C.

The semantics in this PR don't make a distinction between navigations vs subresources vs non-subresource requests (i.e. this PR doesn't involve the [request's destination](https://fetch.spec.whatwg.org/#concept-request-destination) at all), so I chose not to do all the navigation-based variations. The cases I wrote were:

* A with a nested document B. Where B subresources B2 (different origin)
* A with a nested document B. Where B subresources B which redirects to B2

Combined with the rest of the WPT test coverage we have, I think that's a good start. (We already have "A has a nested document B, which subresources B" and similar, I think.)

johannhof reviewed (approved):
This change looks good to me, but I want to be mindful that this could be a big change in web-observable behavior for some embedded third parties. It would change how SAA inherits storage access today in Firefox and Safari, and also how developers got to try it out in our initial 1% experiment in Chrome.

From a Chrome perspective I think we'd like to try and ship this security improvement (pending approval by Blink API Owners etc.), especially given that the old behavior should be easy to restore with SAH. But I'd leave it to @bvandersloot-mozilla and @annevk whether they're comfortable merging this behavior into the spec.

An alternative would be to factor the backwards-incompatible change out of the PR and merge this PR without it.

bvandersloot-mozilla commented:
I think it may make sense to hold off merging this until we are sure the change to only effect the requesting Origin it is something that we can ship for compat reasons. The other part of this is something we should probably do though.

cfredric commented:
I've split this PR into two: #214 adds the request infrastructure so that we can define the current behavior in Chrome/Firefox, and this PR is now stacked on top of #214 and includes the behavior change (`s/site/origin/` in a few places).

I also incorporated @annevk's feedback about using a tri-state instead of a boolean.

cfredric commented:
Since #214 has now been merged, I've rebased this.

annevk reviewed (approved):
This seems good to try. We should be prepared to revert course on this if we run into issues however. We also want to exclude tests for this from Interop 2025.

(You can consider WebKit interested.)

bvandersloot-mozilla reviewed (approved):
Ditto Anne, but s/Webkit/Gecko/

johannhof commented:
So, should we merge this PR, despite not having multiple implementers yet?

annevk commented:
The template in OP should be completed, but I don't see why we wouldn't merge it? To merge you need implementer interest, not concrete implementations.

johannhof commented:
Ok, yeah, I agree, that makes sense. @cfredric can you complete the remaining tasks? Then I'm happy to merge this.

cfredric commented:
Thanks all, I filed bugs and added links to them.
